staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -39,6 +40,583 @@ func RunTestWatch(ctx context.Context, t *testing.T, store storage.Interface) {
+			watchTests: []*testWatchStruct{
+				{basePod("foo6"), false, ""},
+				{basePod("foo7"), false, ""},
+				{basePodUpdated("foo6"), false, ""},
+				{basePodUpdated("foo6"), false, ""},
+				{basePodAssigned("foo6", "bar4"), true, watch.Added},
+			},
+			pred: storage.SelectionPredicate{
+				Label: labels.Everything(),
+				Field: fields.SelectorFromSet(fields.Set{
+					"metadata.name": "foo6",
+					"spec.nodeName": "bar4",
+				}),
+				GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+					pod := obj.(*example.Pod)
+					return nil, fields.Set{"spec.nodeName": pod.Spec.NodeName, "metadata.name": pod.ObjectMeta.Name}, nil
+				},
+				IndexFields: []string{"spec.nodeName"},
+			},
+		},
+		{
+			name:          "cluster-wide watch, request with name, field selecotr with metadata.name, and with spec.nodeName to filter out watch",
+			requestedName: "foo8",
+			watchTests: []*testWatchStruct{
+				{basePod("foo8"), true, watch.Added},
+				{basePod("foo9"), false, ""},
+				{basePodUpdated("foo8"), true, watch.Modified},
+				{basePodAssigned("foo8", "bar5"), true, watch.Deleted},
+				{basePodAssigned("foo9", "bar5"), false, ""},
+			},
+			pred: storage.SelectionPredicate{
+				Label: labels.Everything(),
+				Field: fields.AndSelectors(fields.ParseSelectorOrDie("spec.nodeName!=bar5"), fields.SelectorFromSet(fields.Set{"metadata.name": "foo8"})),
+				GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+					pod := obj.(*example.Pod)
+					return nil, fields.Set{"spec.nodeName": pod.Spec.NodeName, "metadata.name": pod.ObjectMeta.Name}, nil
+				},
+				IndexFields: []string{"spec.nodeName"},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			requestInfo := &genericapirequest.RequestInfo{}
+			requestInfo.Name = tt.requestedName
+			requestInfo.Namespace = ""
+			ctx = genericapirequest.WithRequestInfo(ctx, requestInfo)
+			ctx = genericapirequest.WithNamespace(ctx, "")
+
+			watchKey := "/pods"
+			if tt.requestedName != "" {
+				watchKey += "/" + tt.requestedName
+			}
+
+			w, err := store.Watch(ctx, watchKey, storage.ListOptions{ResourceVersion: "0", Predicate: tt.pred, Recursive: tt.recursive})
+			if err != nil {
+				t.Fatalf("Watch failed: %v", err)
+			}
+			var prevObj *example.Pod
+			for _, watchTest := range tt.watchTests {
+				out := &example.Pod{}
+				key := "pods/" + watchTest.obj.Name
+				err := store.GuaranteedUpdate(ctx, key, out, true, nil, storage.SimpleUpdate(
+					func(runtime.Object) (runtime.Object, error) {
+						obj := watchTest.obj.DeepCopy()
+						return obj, nil
+					}), nil)
+				if err != nil {
+					t.Fatalf("GuaranteedUpdate failed: %v", err)
+				}
+				expectObj := out
+				if watchTest.expectEvent {
+					if watchTest.watchType == watch.Deleted {
+						expectObj = prevObj
+						expectObj.ResourceVersion = out.ResourceVersion
+					}
+					testCheckResultForScopedWatch(t, watchTest.watchType, w, expectObj)
```

Comment:
Continuing my comment - if you will not be watching from RV=0, you should no longer need the for-based checking and will be able to reuse existing functions.

Reply:
Removed the added functions in utils.go.